### ISSUE TITLE
Authentication

### ISSUE THREAD

**Issue by ashleyvega** [vegaprotocol/specs#28]
In GitLab by @edd on Jun 15, 2018, 10:51

# Concepts
## Root account
On Vega we want to be able to associate security keys, collateral from other chains and other disparate pieces of information that will allow us to operate the network. This will consist of at least the following properties:

- A unique ID (Probably a public key, possibly an ethereum public key)
- One or more key pairs
- A password
- One or more public keys for each blockchain that they have associated

## Misc
- Node - a computer running Vega core software. This can be participating in consensus or not (see #78).
- Client - An API client accessing Vega through a node. Currently the trading UI, but anyone can write a client. Is not necessarily on the same physical computer as a node.

# Timeline
## What we want for Nicenet day 1

![Screenshot_2019-01-31_at_12.44.20](/uploads/8946cfbbb725249ee17570debd4fe1a1/Screenshot_2019-01-31_at_12.44.20.png)

- We will be running all nodes on the network
- We will pre-populate a node with a ‘root account’ and key pair for each person we are inviting
- We will assign collateral to each ‘root account’
- Vega will share a password string with that person, which they will enter in to a client to log in to that account
- Transactions committed to the blockchain (orders for example) will be signed with the person’s currently active keypair

![Screenshot_2019-01-31_at_12.44.24](/uploads/ea0c1e1191a9ac6ce73cb0f6cc717723/Screenshot_2019-01-31_at_12.44.24.png)

### Notes
- Multiple key pairs can be associated with a ‘root account’ but in Nicenet day 1, we probably won’t use this
- People will not be able to export or change their Nicenet keys
- People will not be able to associate other blockchain key pairs to their Vega ‘account’
- People will not be able to change their passwords
- The client we direct people to will be the trading UI, hosted by us
- People will also be writing Bots via our API
- The current ‘party string’ we use will be replaced by a UID derived from the public key that the user is currently using
- If a user loses their password, they’re stuck. Although we can just remind them of the password. 
- Some API endpoints will require a user to be authenticated as having access to a keypair

## Near future
![609C1AD7-3E59-4CD1-BE0E-B37C76DB6DA6](/uploads/70f89d53883c01a6141842c5110a1f8c/609C1AD7-3E59-4CD1-BE0E-B37C76DB6DA6.jpeg)
- Users will be able to run nodes (validating or non-validating)
- Users will generate their own key pairs during node setup
- To start move an account to a different node, they key pair will need to be uploaded to the new node.
- Public passive nodes host a read only version of the client for looking at prices, etc. (see #78)

## Further future
- Public "access nodes" may be available to trade without running your own. (Almost certainly not run by us.)
- Users will have to run nodes, or know/work for someone who does. Vega can not and will not.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 15, 2018, 10:51

marked this issue as related to #14

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 15, 2018, 10:58

assigned to @eddhannay

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 15, 2018, 10:59

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 18, 2018, 10:57

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 18, 2018, 10:57

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Dec 4, 2018, 16:20

I had a chat with @futurechimp this morning and he pointed out to me that the concept of a 'user account' and 'accounting' can be separated, and that for nicenet we might be able to survive with 'user accounts' still being strings, as long as the accounting is correct. So for testnet, this would mean:

* We decide that Lionel Hutz is going to be invited to testnet
* Using a new API we assign a balance of `100` to `lionel-hutz-1234`
* We email Lionel a link to the nicenet trading UI
* During onboarding dialog that prompts the user for their name, which is treated just like the current UIDs

Notes:
* All trading/margining is done against that balance
* No additional security checking is done, so if you can guess someone's name, you can steal their account
* Entering a random user string would be fine, you would just most likely be 'logging in' to an account with no balance

Assuming this sounds reasonable, it reduces the amount of thinking we need to do in advance to have user accounts that maintain a constant 'real' balance that users can track their success or lack thereof with. It lets us put off some decisions about what an account really is and how we secure it, especially in a cross chain world

@futurechimp does that roughly sum up what we talked about?
@barnabee @tamlyn10 does that sound like it would be enough for day 1 MVP/nicenet?

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Dec 7, 2018, 14:50

To clarify:
1. This may demonstrate my lack of understanding of how things currently work but am I correct in saying that the trading PnL will 'belong' to the name that lionel adds in during onboarding (let's say "lionel-lion").
2. And then every time lionel reopens Vega, they see their current trading balances as it will open to their name? or will they re-enter their name each time?
3. And then the lionel-hutz-1234 entity's balance will ebb and flow according to lionel's trading performance?

**Comment by ashleyvega** [comment]
In GitLab by @edd on Dec 7, 2018, 14:55

All of that is correct at the moment, yeah - except number 2, it'll remember their name. When we talk about accounting, it's that right now there is no balance associated with lionel's account, just a historical PNL. Building out accounting a bit to track a balance, but leaving the user as it is now is the proposal here.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 11:11

Note: A lot of this post is based on bad assumptions, clarified below

Re-opening this discussion so that we can understand what we're building for Nicenet and beyond. I think this will probably turn in to a meeting, but if @barnabee / @tamlyn10 / @cdm / @futurechimp can pitch in now hopefully we can all get on the same page before we have to organise a meeting.

Note that when I talk about collateral withdrawal, I don't mean steal for their own purposes, I just mean reduce the amount of collateral associated with an account - what happens to the money is out of scope. 

# Where we are now
- A 'user' is a 'party', which is just a string
- Anyone can change their 'user' to any string they want, and the system will act as the new 'user'
- As a result, anyone can place or cancel any order on the system, see anyone's positions, see anyone's collateral
- No collateral is built yet, but if it were anyone could trigger collateral withdrawal any user
- Anyone can create an account just by changing their user string
- _Added after https://gitlab.com/vega-protocol/product/issues/28#note_136673765_: users run a front-end client connecting to a node that we run

# Where we want to be in a year+
- Only the user can place or cancel an order as that user
- However it's fine that anyone can, via the API at least, see the collateral balance or current positions for any user on the system
- Only a user can withdraw collateral for their own account
- 2-factor authentication
- New users can create an account
- _Added after https://gitlab.com/vega-protocol/product/issues/28#note_136673765_: Users run a front-end client connecting to a Vega node that anyone can be running, including on their local machine

# Where we want to be for Nicenet
![IMG_0038](/uploads/d69bb091a051c85f93c9f4054bddf534/IMG_0038.jpg)
- We want to be able to create a user and give access to that user to someone else
- We want to to be able to assign collateral to that account
- _Added after https://gitlab.com/vega-protocol/product/issues/28#note_136673765_: Users run a front-end client connecting to a Vega node on their local machine

# Things I don't understand
- Initially we talked about using Metamask/Brave/any-old-ethereum-wallet-in-the-browser to use an existing Ethereum key pair to sign a transactions on Vega. That doesn't seem like an ideal plan for a multi-blockchain world, so is this still something we're keeping on the cards? If so, presumably after nicenet?
- GraphQL roughly supports the same header/auth schemes as REST, which is great. I assume if we're using session cookies/JWT/whatever, it's an auth session with one particular Vega node~, and switching nodes would mean re-authing?~ (_addressed in https://gitlab.com/vega-protocol/product/issues/28#note_136675760_)
- I could list our API endpoints that require auth (basically: all GQL mutations) but presumably we can't rely on API auth alone (_Addressed in https://gitlab.com/vega-protocol/product/issues/28#note_136671354_)


--- 
There are a whole host of security concerns that are related, but not entirely covered here. I'll just quote @ashleyvega's slack message and we can make new tickets for aspects that are going to need to be covered in the coming months:

> Ashley [10:00] @eddhannay  Auth is a large and smelly kettle of fish. I'll definitely be waxing my oar, ready to stick it in, regarding: signups, logins, IP addr tracking, tokens, multiple simultaneous logins allowed, 2FA/OTP, multi-2FA (like Kraken), logout timeouts, security questions, password resets, API access tokens, etc).

> and also http/https (hint: one of those is baaad)

> Ashley [10:02] surely (*surely*!) it can't be that hard to plug in Google Authenticator. it's not only that we're in a race to get to nicenet asap, we should also consider what the nicenet users will think of the system and its security. At the least, let's have a 2FA tab in the user's settings page which says "Not enabled for nicenet".

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 15:17

Things to think about - note I am not sure the web3 login way is the best way to go as you will see, but it's definitely not off the table...:

1. A **user** is probably different to an identity (keypair) that holds collateral on say Ethereum. So I think the Vega node will need to generate or import a keypair for the user that will be used to sign Vega transactions *and* will eventually need a way to link one or more collateral wallets by proving ownership (e.g. with Metamask). I believe this proof of ownership will either need to be persisted on the Vega blockchain via a TX *or* every TX impacting external addresses e.g. withdrawals will need to be signed by both the Vega private key and the appropriate [Ethereum] private key.

1. The Vega node would store the Vega keypair along with the relevant associations and details, encrypted with a password/phrase (+[option of] 2FA) and that is what would be used to login. This means that a session token/whatever provided by the node would be enough for the Vega node to sign and submit a TX on Vega (unless we wanted to require an additional challenge e.g. the current 2FA code on those operations - probably...). NB: this is how Ethereum (e.g. Geth, Parity), Bitcoin, et al. work.

1. We may not want the Vega node to store private keys for other chains so could use MetaMask/whatever for authing transactions on those chains (e.g. withdrawal) and only store the association and public key on Vega.

1. For Nicenet we have no integration so we can just do [2] above and store (encrypted) Vega keypair on the node, and inmplement whatever login we want with that (e.g. the authentication doesn't really need to change when you introduce Ethereum integration, only need to add the way of associating the Eth wallet via Metamask or whatever).

1. We could even get away for Nicenet for a while with storing a randomly generated ID on the node like the client currently generates (rather than a keypair) but that won't work for long (anyone can still fake a transaction by anyone else) and seems like unecessarily kicking the can down the road.

1. Vega will absolutely need to support multiple accounts/identities/wallets being stored and used on a node. It's tempting to try not do so for Nicenet, but this is really only a data model problem and shouldn't be horrendously complex as the node already needs to know the full network state after all and CQRS means we're just querying read stores for different IDs, so I think it's highly likely we should start out with multi-wallet/identity/account support on the nodes.

1. At least initially there are no security questions, password resets, etc. - you are creating a keypair and connecting to a decentralised network. We don't maintain a database of those things (and can't). Any such centralised stuff would be an additional layer and service on top of Vega and I don't know if we will ever operate such a thing (it's legally difficult, for a start!) (NB: that said, we may actually create the keypairs on nodes we run for Nicenet users and give them the 'login' details to unlock them, but that will likely not be sustainable for Mainnet or even a more production-like Testnet.)


On Edd's questions:

> Initially we talked about using Metamask/Brave/any-old-ethereum-wallet-in-the-browser to use an existing Ethereum key pair to sign a transactions on Vega. That doesn't seem like an ideal plan for a multi-blockchain world, so is this still something we're keeping on the cards? If so, presumably after nicenet?

It's needed to associate a wallet as per above but I think we probably just make and use our own key stuff for Nicenet and Vega wallets/IDs in general. There is definitely the option of instead forcing the 'root' ID to be one on Ethereum (or potentially any of our supported blockchains) rather than code our own, then we'd use Metamask for all login I guess, but I'm not sure that's the right approach even though it avoids writing as much crypto code and is where we started in the description of the original ticket. @futurechimp may have opinions here...

> GraphQL roughly supports the same header/auth schemes as REST, which is great. I assume if we're using session cookies/JWT/whatever, it's an auth session with one particular Vega node, and switching nodes would mean re-authing?

Switching nodes definitely needs reauthing, and importing your keypair etc. You don't want every node holding your private key.

> I could list our API endpoints that require auth (basically: all GQL mutations) but presumably we can't rely on API auth alone

Assuming API auth 'unlocks' your keypair I don't see why not? The client should never see or use private keys really IMO. As mentioned above we may want to add a challenge to anything that posts certain types of transaction to see the password again and/or a 2FA code. API auth may also be used to store/retrieve settings or see 'my' orders and collateral although that's not strictly necessary as long as we have an API to retrieve 'my' public key/ID.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 15:43

@barnabee thanks for the long reply, I'm taking it all in now

> Assuming API auth 'unlocks' your keypair I don't see why not? The client should never see or use private keys really IMO. 

My thinking here was that the Vega APIs are currently the convenient way to write valid transactions to the blockchain, but we are also operating a public blockchain that someone could write whatever they want to. If all operations are going to be signed in the Vega application, and unsigned operations ignored, then this isn't a problem vector.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 15:56

changed title from **Authentication{- / web3-}** to **Authentication**

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 17:08

@barnabee I’ve been trying to turn this in to a basic set of requirements for the onboarding UI, but:

> So I think the Vega node will need to generate or import a keypair

I think I need some clarification on where the keys come from.

> The client should never see or use private keys really IMO

Is this ‘with the exception of when the user first connects to a node’ or never never? I assume it’s the former - a keypair is created on my local machine, and I send that to a node which creates a ‘root id’ and associates the ‘root id’ with my new keypair.

If the node generates the keypair and doesn’t give me access to it, that feels like it has a whole host of issues - and also I can’t: 

> Switching nodes definitely needs reauthing, and importing your keypair etc. You don't want every node holding your private key.

Without an API to download my keypair, which seems problematic in a situation where let’s say a node goes offline and I want to withdraw collateral or

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 18:07

> My thinking here was that the Vega APIs are currently the convenient way to write valid transactions to the blockchain, but we are also operating a public blockchain that someone could write whatever they want to. If all operations are going to be signed in the Vega application, and unsigned operations ignored, then this isn't a problem vector.

Yeah, I think it's perfectly valid that people can write what they want to the blockchain, and indeed unstoppable, and there are a few ways they could potentially do this (not that we have to support them all):

- via an authed API call, where the Vega node signs the TX
- via an un-authed API call, where a pre-signed TX is submitted
- from the command line calling Vega, with creds to 'unlock' the encryption key for a wallet
- from the command line calling Vega, passing a signed TX
- by connecting their own client software and passing a signed TX

As I think you alluded, if we assume all TXs must be signed and the Vega node software will do that if you have an authed session, I think it's all OK.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 18:14

> I think I need some clarification on where the keys come from.

They are generated by the node, possibly doable from the front end but if not it would be a command line e.g. something like `vega --keygen --name edd --password 8008135 > edd.dat`

> Is this ‘with the exception of when the user first connects to a node’ or never never? I assume it’s the former - a keypair is created on my local machine, and I send that to a node which creates a ‘root id’ and associates the ‘root id’ with my new keypair.

My assumption is that node and local machine are the same. That's how it works for Bitcoin, Ethereum, etc. This doesn't mean you have to run a validator node but you should be using the Vega software to create a keypair, and most likely to host the API instance you're using, etc. It seems very unlikely that "there are *n* nodes and every trader connects to one of them to run the UI" will be supported, at least initially. I think people will need to be running nodes *after Nicenet*.

> If the node generates the keypair and doesn’t give me access to it, that feels like it has a whole host of issues - and also I can’t:

See above. You generate the keypair on your node. Again, for Nicenet we may do it for you and send the login stuff to you but that's more because we want to keep the network simple and in our control, and with no real money it doesn't matter.

> Without an API to download my keypair, which seems problematic in a situation where let’s say a node goes offline and I want to withdraw collateral or

I think this is covered by the above. The kaypair is on your machine that runs the node.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 18:19

**Note on all the above:** this all needs to be checked for archtiectural soundness, etc. I am assuming it is ok to have lots of clients that are read only join the network in addition to the (much smaller) number of validators and we need to test this, and may not actually do this until after Nicenet (with the workaround that we do the keygen for everyone on our small number of nodes, and give them the API logins to start with). However it doesn't seem too crazy an assumption because it's consistent with other blockchains, I don't see how else it could realistically work, and Tendermint docs contain pages like https://tendermint.com/docs/tendermint-core/light-client-protocol.html which at least suggest they are thinking about clients as well as core validators.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 18:28

Note: this comment is all wrong. See below.

~~Aha! The One True Whiteboard still has "other people running nodes?" on it, which reconfirmed my belief that in Nicenet we'd be running the nodes and people would just be visiting a URL.~~

~~So that reopens https://gitlab.com/vega-protocol/product/issues/20, but makes things much simpler - I have to do less weird mental gymnastics when reading your first big post. Phew! Alright. I'm going to add that as a note on my first big post.~~

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Jan 30, 2019, 18:30

`vega --keygen --name edd --password 8008135 > edd.dat`

^ Liking this and my new CLI stuff makes this elegant to add to the project.

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Jan 30, 2019, 18:33

Not sure if this helps here but `other people running nodes?` was discussed by me and @tamlyn10 last Friday and I think that relates to read only nodes not participating in concensus/trading. Which does open a question about auth on those kind of nodes. Scope wise, I think we'd need to check if these read nodes are part of nice-net.

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 18:40

Read-only clients as a distinct from light clients? Regardless, I don't believe either is in scope for nicenet.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 18:42

a) @cdm is correct, we are absolutely talking about read only, 'part time' nodes. 

b) I tried to make clear that **for Nicenet other people running nodes is not necessary, not even for read only nodes** - they never need to see their keypair even, we just tell them which node has it and what their pubkey address (which ~= user ID) and the password to access that keypair are and they are good to go. Hell we can even live without actually signing TXs to start with I suspect, if we really have to.

c) 'Read only' clients can be light or full, I don't think validators can be light clients.

Fun fact: they aren't completely read only, they can still submit TXs to the network but they are read only from the point of view of consensus: they don't participate in the consensus algorithm, hold stake, get rewards, vote for or propose blocks, etc. ... they just listen to the blocks that get produced, run the Vega code on the result, serve up the APIs, and forward valid submitted transactions to the validators.

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Jan 30, 2019, 18:52

I've created parent product ticket stub #78 as the place for all truth on that matter, please fill it out with additional info :pen_ballpoint:

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 30, 2019, 18:55

Also, other people running [passive/read only] nodes will definitely come in pretty quick but the scope question mark came about because we had done no investigation whatsoever into how easy it would be to get such a node connected into Tendermint (and a bit of #20 but initially the numbers will be small enough for us to handhold), and so we weren't sure whether it should be in scope, as you can potentially ignore loads of (most) auth stuff if you can assume every user is running their own passive node for Nicenet, so in some ways it's appealing...

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 30, 2019, 19:00

Alright - I think I'm up to speed now, so I've added a note to my flagrantly wrong comments above and will write out a more correct version, which I'll bump up to be the actual ticket if it's approved. Thanks for going through all of this, I think I had a mix of disparate conversations & old decisions in my head, so it's great that everyone has taken the time to discuss this somewhere it's easy to look back over

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 12:45

# Concepts
## Root account
On Vega we want to be able to associate security keys, collateral from other chains and other disparate pieces of information that will allow us to operate the network. This will consist of at least the following properties:

- A unique ID (Possibly an ethereum public key)
- One or more key pairs
- A password
- One or more public keys for each blockchain that they have associated

## Misc
- Node - a computer running Vega core software. This can be participating in consensus or not (see #78).
- Client - An API client accessing Vega through a node. Currently the trading UI, but anyone can write a client. Is not necessarily on the same physical computer as a node.

# Timeline
## What we want for Nicenet day 1

![Screenshot_2019-01-31_at_12.44.20](/uploads/8946cfbbb725249ee17570debd4fe1a1/Screenshot_2019-01-31_at_12.44.20.png)

- We will be running all nodes on the network
- We will pre-populate a node with a ‘root account’ and key pair for each person we are inviting
- We will assign collateral to each ‘root account’
- Vega will share a password string with that person, which they will enter in to a client to log in to that account
- Transactions committed to the blockchain (orders for example) will be signed with the person’s currently active keypair

![Screenshot_2019-01-31_at_12.44.24](/uploads/ea0c1e1191a9ac6ce73cb0f6cc717723/Screenshot_2019-01-31_at_12.44.24.png)

### Notes
- Multiple key pairs can be associated with a ‘root account’ but in Nicenet day 1, we probably won’t use this
- People will not be able to export or change their Nicenet keys
- People will not be able to associate other blockchain key pairs to their Vega ‘account’
- People will not be able to change their passwords
- The client we direct people too will be the trading UI, hosted by us
- The current ‘party string’ we use will be replaced by a UID derived from the public key that the user is currently using
- If a user loses their password, they’re stuck. Although we can just remind them of the password. 
- Some API endpoints will require a user to be authenticated as having access to a keypair

## Near future
- Users will be able to run nodes (validating or non-validating)
- Users will generate their own key pairs during node setup
- To start move an account to a different node, they key pair will need to be uploaded to the new node.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 31, 2019, 12:55

Looks good!

> A unique ID (Possibly an ethereum public key)

And in any case very likely the public key from some sort of keypair.

> The client we direct people too will be the trading UI, hosted by us

And the APIs. Some of the very first Nicenet users will want to write bots (and we want them to.)

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Jan 31, 2019, 12:56

Also:

## Near-ish future

- Traders will have to be running a node
- Public passive nodes host a read only version of the client for looking at prices, etc.

## Further future

- Public "access nodes" may be available to trade without running your own. (Almost certainly not run by us.)

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 13:02

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 13:03

I've boosted this to the top and am including your edits - thanks!

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 13:04

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 13:24

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jan 31, 2019, 13:32

marked this issue as related to #72

**Comment by ashleyvega** [comment]
In GitLab by @edd on Apr 17, 2019, 11:05

marked this issue as related to #105

**Comment by ashleyvega** [comment]
In GitLab by @mattrussell36 on Apr 29, 2019, 19:13

@jeremyletang worth reading through the discussion here. I think this covers off some of the points we discussed?

@barnabee ... I think the answer is no. But could you confirm if the token returned by the node should expire?

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Apr 29, 2019, 19:59

@mattrussell36 for Nicenet, no, but I suspect that in future (a) the token shouldn’t survive a restart of the node; and (b) the session should time out and forget the token and any decrypted private keys after inactivity, so that a long running background node poses less of a security risk if someone obtains access to the machine.

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 16, 2019, 17:11

marked this issue as related to trading-core#287

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 16, 2019, 17:12

removed the relation with #105

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 16, 2019, 18:48

marked this issue as related to trading-core#289

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 16, 2019, 18:57

marked this issue as related to trading-core#290

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 11:32

@barnabee I'm writing up tickets for our proposed testnet auth solution at the moment and linking them to this ticket (trading-core#290, trading-core#288, trading-core#287) - one important bit that's dropped from this ticket initially is signing & verifying transactions. Clients will still auth with a token, but transactions won't be signed so that we have more time (and more time with product) to work through some thoughts, specifically:

- Key publishing
- When/where to do signature verification without slowing everything down too much

Let me know if that's an issue. See #111 for the things we need to thrash out before starting on signing.

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 12:13

marked this issue as related to trading-core#291

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 14:32

marked this issue as related to client#347

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 14:38

marked this issue as related to client#348

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 14:46

marked this issue as related to #111

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 17, 2019, 16:04

marked this issue as related to client#327

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 22, 2019, 14:25

marked this issue as related to devops-infra#21

**Comment by ashleyvega** [comment]
In GitLab by @edd on Jun 12, 2019, 14:07

closed

**Comment by ashleyvega** [comment]
In GitLab by @ashleyvega on Jul 19, 2019, 13:10

marked this issue as related to #124

**Comment by ashleyvega** [comment]
In GitLab by @ashleyvega on Aug 13, 2019, 13:33

mentioned in merge request !13
